src/udp_wrap.h

```diff
@@ -30,10 +30,94 @@
+
+  // Called right before data is received from the socket. Must return a
+  // buffer suitable for reading data into, that is then passed to OnRecv.
+  virtual uv_buf_t OnAlloc(size_t suggested_size) = 0;
+
+  // Called right after data is received from the socket, and includes
+  // information about the source address. If `nread` is negative, an error
+  // has occurred, and it represents a libuv error code.
+  virtual void OnRecv(ssize_t nread,
+                      const uv_buf_t& buf,
+                      const sockaddr* addr,
+                      unsigned int flags) = 0;
+
+  // Called when an asynchronous request for writing data is created.
+  // The `msg_size` value contains the total size of the data to be sent,
+  // but may be ignored by the implementation of this Method.
+  // The return value is later passed to OnSendDone.
+  virtual ReqWrap<uv_udp_send_t>* CreateSendWrap(size_t msg_size) = 0;
+
+  // Called when an asynchronous request for writing data has finished.
+  // If status is negative, an error has occurred, and it represents a libuv
+  // error code.
+  virtual void OnSendDone(ReqWrap<uv_udp_send_t>* wrap, int status) = 0;
+
+  // Optional callback that is called after the socket has been bound.
+  virtual void OnAfterBind() {}
+
+  inline UDPWrapBase* udp() const { return wrap_; }
+
+ protected:
+  UDPWrapBase* wrap_ = nullptr;
+
+  friend class UDPWrapBase;
+};
+
+class UDPWrapBase {
+ public:
+  virtual ~UDPWrapBase();
+
+  // Start emitting OnAlloc() + OnRecv() events on the listener.
+  virtual int RecvStart() = 0;
+
+  // Stop emitting OnAlloc() + OnRecv() events on the listener.
+  virtual int RecvStop() = 0;
+
+  // Send a chunk of data over this socket. This may call CreateSendWrap()
+  // on the listener if an async transmission is necessary.
+  virtual ssize_t Send(uv_buf_t* bufs,
+                       size_t nbufs,
+                       const sockaddr* addr) = 0;
+
+  // Stores the sockaddr for the peer in `name`.
+  virtual int GetPeerName(sockaddr* name, int* namelen) = 0;
+
+  // Stores the sockaddr for the local socket in `name`.
+  virtual int GetSockName(sockaddr* name, int* namelen) = 0;
+
+  // Returns an AsyncWrap object with the same lifetime as this object.
+  virtual AsyncWrap* GetAsyncWrap() = 0;
+
+  void set_listener(UDPListener* listener);
+  UDPListener* listener() const;
+
+  static UDPWrapBase* FromObject(v8::Local<v8::Object> obj);
+
+  static void RecvStart(const v8::FunctionCallbackInfo<v8::Value>& args);
+  static void RecvStop(const v8::FunctionCallbackInfo<v8::Value>& args);
+  static void AddMethods(Environment* env, v8::Local<v8::FunctionTemplate> t);
+
+ private:
+  UDPListener* listener_ = nullptr;
+};
+
+class UDPWrap final : public HandleWrap,
+                      public UDPWrapBase,
+                      public UDPListener {
+ public:
+  enum InternalFields {
+    kUDPWrapBaseField = HandleWrap::kInternalFieldCount,
+    kInternalFieldCount
+  };
```

Comment:
This works for now, but keep in mind that the purpose of this was to introduce another subclass of `UDPWrapBase` in the QUIC PR… in that case, we may want to move the enum here to `UDPWrapBase`?

Reply:
Yep, let's worry about that one minor change then